inform-db/src/test/java/uk/ac/ucl/rits/inform/informdb/TestCopyConstructor.java

```diff
@@ -0,0 +1,259 @@
+                return STRINGS.get(index);
+            case "uk.ac.ucl.rits.inform.informdb.TemporalFrom":
+                return new TemporalFrom(Instant.now(), Instant.now().plus(1, ChronoUnit.SECONDS));
+            case "java.time.LocalDate":
+                return LocalDate.now();
+            case "java.util.List":
+                return List.of();
+        }
+
+        throw new StringIndexOutOfBoundsException("Did not find " + className + " in base types");
+    }
+
+    /**
+     * Create a new instance of a class given a name of it
+     *
+     * @param className Name of a class
+     * @return Instance of the class
+     */
+    Object newInstance(String className){
+
+        incrementIndex();
+
+        if (BASE_ClASS_NAMES.contains(className)){
+            return newBaseInstance(className);
+        }
+
+        Class<?> entity = classFromName(className);
+        Object instance = defaultConstructedInstance(entity);
+        setAllFieldsOf(entity, instance);
+
+        return instance;
+    }
+
+    /**
+     * Given an instance of a class that corresponds to a particular entity set all the fields/attributes with
+     * default values which may require instanciating new instances of wither base classes (e.g. integers or strings,
+     * or recursively calling newInstance and setting fields of those parameters)
+     * @param entity Class entity
+     * @param instance Class instance
+     */
+    void setAllFieldsOf(Class<?> entity, Object instance){
+
+        for (Method method : setterMethodsOf(entity)){
+
+            Object[] params = Arrays.stream(method.getParameterTypes())
+                    .map(p -> newInstance(p.getName()))
+                    .toArray();
+
+            try {
+                method.invoke(instance, params);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Is a method of an entity a copy constructor?
+     */
+    boolean isCopyConstructorOf(Constructor<?> constructor, Class<?> entity) {
+        return constructor.getParameterCount() == 1
+                && constructor.getParameterTypes()[0].getName().equals(entity.getTypeName());
+    }
+
+    /**
+     * Create a copy of a class instance given its entity and current instance by invoking the copy constructor
+     * @param entity Class entity
+     * @param instance Class instance
+     */
+    Object copyOf(Class<?> entity, Object instance){
+
+        Method copyMethod = Arrays.stream(entity.getMethods())
+                .filter(m -> m.getName().equals("copy"))
+                .findFirst().orElseThrow(() -> new RuntimeException("No copy method"));
+
+        try {
+            return copyMethod.invoke(instance);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+            e.printStackTrace();
+            throw new RuntimeException("Failed to call copy method");
```

Comment:
probably better to just throw the exception from the existing exception 
```suggestion
            throw new RuntimeException("Failed to call copy method", e);
```